### ISSUE TITLE
Chimp 2.0 Proposal

### ISSUE THREAD

**Issue by samhatoum** [xolvio/chimp#614]
# Chimp 2.0
Chimp was designed to help users ***deliver higher quality software faster through testing***, which is also Xolv.io's mission as a consulting company. Chimp did this so by providing users with:

1. An easy to use synchronous API based on fibers
2. An easy installation and configuration of the testing stack
3. Fast feedback using smarts like the innovative watch mode and session management
4. First class Meteor support

The way in which Chimp did the above was to utilize existing frameworks and libraries.

It seems that @christian-bromann has been doing a fantastic job with the [WebdriverIO test-runner](http://webdriver.io/guide/testrunner/gettingstarted.html), and from what we can see steps 1 and 2 from above are taken care of, and there is a watch mode that takes care of some of step 3 (WebdriverIO does not run when non-test files changes whereas Chimp does).

It seems there's a lot of duplication between the test-runner portion of Chimp and WebdriverIO. However, we think there is still a need to help users ***deliver higher quality software faster through testing***.

After some brainstorming with  @lgandecki - the founder of The Brain, a software development shop with a high emphasis on testing and a maintainer of Chimp - we've decided to rewrite Chimp to use a new set of libraries to work towards our respective company ethos.

We thinking to use [Gulp](gulpjs.com) as the "base" platform since it is far more than just a test runner, it's a full build system. In fact, you can see that we already started to use Gulp in our [Quality Faster](https://github.com/xolvio/qualityfaster/blob/dev/gulpfile.js) repo, so this seems like a natural progression.

What Chimp then becomes an overall package that is comprised of Gulp itself and a set of Gulp tasks that allow you to do the full lifecycle of testing using an opinionated testing stack at every testing level:

* **Client Unit Testing**:
  `gulp` + `karma` + `mocha` + `chai` + `td` + `chimp-karma-watch`
* **Server Unit Testing**:
  `gulp` + `mocha` + `chai` + `td` + `chimp-mocha-watch`
* **Domain  Testing with CucumberJS**:
  `gulp` + `cucumberjs` + `chai` + `td` + `chimp-cucumber-watch`
* **End-to-end Testing with Cucumber**:
  `gulp` + `webdriverio` + `chimp-cucumber-webdriverio-watch-plugin` - `chimp-meteor-webdriverio-watch-plugin`
* **End-to-end  Testing with Mocha**:
  `gulp` + `webdriverio` + `chimp-mocha-webdriverio-watch-plugin` - `chimp-meteor-webdriverio-watch-plugin`

> NOTE 1: Each testing level also comes with the watch mode you know and love
> NOTE 2: Package names are not exact, but they illustrate the approach

Since we're using Gulp, it means users can also extend the stack with tasks like this for example:
* Start Meteor Application
  `gulp` + `xolvio-meteor-gulp`

We are basically going to modularize the heck out of Chimp, and as other frameworks do what Chimp does (which will invariably happen), we can reuse their code and contribute rather than try to compete.

### Backwards Compatibility
We will do our best to be backward compatible, but where there are design challenges or better best practices to follow we will provide a migration guide for those that wish to migrate their tests over.

The aim of this initiative is to become more friendly with the existing OSS ecosystem and for Chimp to become a set of opinions on ultimately helping teams deliver higher **Quality, Faster**.

Thoughts / opinions / suggestions / feedback all very welcome.

**Comment by christian-bromann** [comment]
> It seems that @christian-bromann has been doing a fantastic job with the wdio test-runner,

Thanks

> wdio does not run when non-test files changes whereas Chimp does

Happy to enhance the current watch feature to do that too. As long as the feature is not too much application focused I am happy to add it to the WebdriverIO test runner so that Chimp can concentrate on its main goal.

**Comment by samhatoum** [comment]
Awesome @christian-bromann. We were already thinking about a few PR's we'd send your way to help. I'll post a separate issues on the WebdriverIO repo as soon as we get down to some designs.

Thanks for your amazing work so far!

**Comment by maxmckenzie** [comment]
Nooo please not Gulp. This locks me into having to use it. I can run tests just fine from node package scripts. 

> I hope I understood correctly, that this would requier me to implement gulp in my projects

The thing I like the most about chimp is that it is based on an npm command and a config. This means that I can use anything webpack, gulp, some custom creation from a lunatic. i just add some package scripts and boom. no grunt, no gulp no nothing.

Not only that but it also means I can sneak it into a project without having to go through all the stakeholders. Adding an npm script to a PHP, node, python app etc is easy enough and most people will be fine with it and agree.

But if I have to stand there and try to convince the stakeholders to move the whole build system into gulp. That would be knocked outa the sky so fast, it'd hurt

**Comment by lgandecki** [comment]
@christian-bromann Thanks for joining the discussion!

We would like to have two things:
1) rerun tests when step_definitions change. 
We understand the difficulty here - when spec file change, you reload only that spec file. Easy.
When you change step_definition, it's not so clear which specs to reload. And reloading all is not a good idea. ;-) 
Our approach in chimp was to have the features you are currently working on tagged with @watch , so whenever step_definitions are changed, all tagged specs are reloaded. 

Long term, we can add an option to automatically reload specs based on their usage of step_definitions. 

2) Rerun tests on Meteor restart - we were getting the information from the server and reloading the @watch tagged tests.
I think that as long as we have access to this.framework.run from plugins (haven't researched this enough yet), we should be able to achieve both tasks. (and actually, other people might be interested in writing custom "reloaders plugins" for their apps, once we show them how to get this done)

**Comment by samhatoum** [comment]
@maxmckenzie I see what you mean. 

From an installation point of view I think we can still have `npm i chimp` and it should work similarly to how it works today. The difference being it would use Gulp / wdio under the hood instead of the proprietary approach we use today.

However, if you are using Gulp yourself then you can just use the packages like `chimp-end-to-end` in a decoupled fashion.

So you don't have to move to using Gulp as a build system because Chimp is using it under the hood. It just makes the test runner part of chimp much more powerful as it will be standing on the shoulders of Gulp/wdio.

Does that answer your question / address your concern?

**Comment by maxmckenzie** [comment]
Yes, it does. Thank you (panic averted)

So if I understand correctly. At the moment I just have chimp as the command to run the tests and a chimp-config.js that manages all my mocha, wdio, cucumber configs. I also just have chimp in my package.json

Will this will still be the case?

In terms of how chimp does it, somewhere in my node_modules, it will use gulp.

> basically provided I never have to see the word gulp I'm fine :)

**Comment by maxmckenzie** [comment]
In regards to only using parts of the test runner e.g just chimp-cucumber or just chimp-mocha etc is there a benefit to that? If I run just gherkin and step definitions it does not seem to leverage any other frameworks to do it? If i do then its already there which is super usful

An example of how I do this may be an idea 
https://github.com/gigya/node-gigya-boilerplate

> still actively being built.

note that I run mocha stuff but the only dependency is chimp. That's a sweet feature

basically, chimp becomes the gatekeeper to the other frameworks and makes it easier to work with.

**Comment by lgandecki** [comment]
That's the idea. As Sam said, we will try to make it as backward compatible as possible. If there are some things from the chimp-config.js that we won't be able to directly support we will provide a migration guide. We will know more in the next few weeks.

**Comment by maxmckenzie** [comment]
Cool. I mean i dont even use a config :D thats how sweet chimp is at sorting stuff out

**Comment by lgandecki** [comment]
right, if you don't use a config, I'm pretty sure you will be able to just upgrade. The only thing I'm not so sure yet of is the multibrowser option (that we heavily use for testing reactivity, so I will personally put a lot of effort to have this working one way or the other)

**Comment by erwinheitzman** [comment]
It would be great to be able to set the packages that you want to use by specifying them in the chimp config like so:
```
{
  packages: { wdio: true, cucumber: true }
}
```
And then have a function that installs these packages using npm so that we do not end up with packages that aren't needed.

The above could install the latest (stable?) versions and the perhaps allow the versions to be set in the config as well like so:
```
{
  packages: { wdio: "4.4.0", cucumber: "2.0" }
}
```

**Comment by christian-bromann** [comment]
Let's make sure to not confuse `wdio` with `webdriverio`. I would rather use the longer string in order to be more explicit about what is actually used.

**Comment by samhatoum** [comment]
Correct me if I'm wrong @christian-bromann:

WebdriverIO = the bindings
wdio = WebdriverIO test runner

Did I get that right? Or is WebdriverIO the overall project, and within that there are bindings and a runner?

**Comment by christian-bromann** [comment]
WebdriverIO is the overall project. `wdio` is the cli command and I sometimes refer to it because it is shorter to write. I just want to avoid that someone is looking for `wdio` on NPM and finds [this](https://www.npmjs.com/package/wdio). And yes WebdriverIO contains a testrunner out of the box.

**Comment by samhatoum** [comment]
All references have now been changed to WebdriverIO/webdriverio

**Comment by tiagolr** [comment]
I don't know what are the benefits of using gulp under the hood, but if what is good now remains functional, all is good.

Tried wdio and while it seems a solid project it lacks some things: watch mode has problems (nothing happens when I change features), tag arguments don't work, cucumber json output is not supported, opens a process for every feature even if they are ignored using tags, using `~@ignore` tag causes problems (blank output), forces to use one process per feature, only outputs after feature tests are done... These are small problems that can probably be fixed easily, but they are there..

For that it would be great to still have Chimp as a reliable test-runner, with updated cucumber and other libs, innovative features are also welcome, Chimp would also benefit from a cleaner codebase and tests.

Cheers,

**Comment by christian-bromann** [comment]
> cucumber json output is not supported

WebdriverIO doesn't support native cucumber reporter but provides an own set of [reporter](http://webdriver.io/guide/reporters/json.html). Note that unit test reporter don't work when running e2e tests due to different requirements.

> opens a process for every feature even if they are ignored using tags

This is a know issue and we're thinking about how to solve this issue. There are workarounds for this though.

Since I am not using Cucumber on daily basis the integration might be a bit behind. Happy to find someone who can help me to improve this.

**Comment by tiagolr** [comment]
@christian-bromann glad for your reply, one quick improvement if I noticed correctly, is that the output is only sent to the stdio after a suite finishes.

A quick fix for this is set the stdio from child processes to inherit the parent: 
```
childProcess.exec(command, { stdio: 'inherit' }, function(error) { ... }
```
then the output from tests is printed in real-time. 

BR,

**Comment by capaj** [comment]
I don't think gulp is a wise choice for a testing framework. Gulp was created with a build tool in mind. Test framework needs can be addressed with it, but there will always be ton of unused code and just fighting gulp APIs to get it to do what the test framework wants to do.

**Comment by samhatoum** [comment]
> @tiagolr: I don't know what are the benefits of using gulp under the hood

> @capaj I don't think gulp is a wise choice for a testing framework. Gulp was created with a build tool in mind. 

I appreciate your comments guys, let me explain "why gulp?".

Chimp is a test runner but not a testing framework per se. It uses other testing frameworks like Mocha/Jasmine/Cucumber and configures the tooling required like Chai, WebdriverIO's selenium bindings and so on.

WebdriverIO now sports a test runner that does mostly what Chimp does, and with a bit of effort we can work together and get the best of Chimp's *test runner* features in there.

However, developers don't (shouldn't) just want to run end-to-end tests. TDD'ers and BDD'ers want to test their *whole application*, and that is an overall workflow concern and not just a test runner. In my experience connecting all the toolset for the entire workflow is a struggle - especially for newcomers, and having an opinionated full testing stack in the JS testing space seems to be lacking.

Chimp already started going down this road with the **domain testing** mode, whereby the watch mode was applied only to Cucumber without the browser portion. [See here for more details about the reason for this testing mode](https://github.com/xolvio/qualityfaster/blob/6f4199b2a526a049953dc1b554d1ae7cf482df10/content/TESTING-SUCCESS-FACTORS.md#lesson-4-do-not-test-the-domain-through-the-ui)

And so testing the whole application requires the following:
* client
* server
* end-to-end through the web/browser
* end-to-end through API's

I postulate that a large part of testing the whole application is a workflow concern and here is the purpose of gulp:<br/>*(taken from [gulpjs.com](gulpjs.com))*

> gulp is a toolkit for automating painful or time-consuming tasks in your development workflow

Here's is a list of the test runners that I know and what they're good at:

* WallabyJS - server / client - real-time local - no CI
* Karma - client only - watch mode local - CI support
* WebdriverIO - browser-to-end - watch mode local - CI support
* Mocha - server/client - watch mode local - CI support

Jasmine / Cucumber don't have a watch mode of their own, which is where Chimp / WebdriverIO currently fill the gap for end-to-end testing.

Now let's say you want to do something a little more involved like collect test coverage reports from all the testing framework runs to create a combined test coverage report. Or you want to run all the tests in parallel locally and on CI to save time. All these things are workflow management concerns that GulpJs does a great job of and this is the direction I'm wanting to take Chimp in. Not just an end-to-end test runner, but an opinionated testing workflow and testing stack to help you build high quality at speed.

At Xolv.io make a lot of such choices with clients to hyper-optimise their testing stack. [See this article: How to Design and Build a Hyper-Fast Test Automation Stack](https://medium.com/quality-functions/how-to-design-and-build-a-hyper-fast-test-automation-stack-84dc6985e569)

So my thinking is that Chimp can focus on providing an hyper-optimised testing stack rather than just being one end-to-end testing framework. It can utilize the best of the frameworks and testing tools out there do this job.

Thoughts?

**Comment by samhatoum** [comment]
Chimp 2.0 is well underway. closing this issue

**Comment by Windso** [comment]
Any update?
